src/go/rpk/pkg/auth0/auth0.go

```diff
@@ -0,0 +1,218 @@
+	defer ticker.Stop()
+	for range ticker.C {
+		token, err := cl.getToken(ctx, code.DeviceCode)
+		if err == nil {
+			return token.AccessToken, token.ExpiresIn, nil
+		}
+		if errors.As(err, &token) {
+			switch token.Err {
+			case "authorization_pending",
+				"slow_down":
+			default:
+				return "", 0, err
+			}
+		}
+		if time.Now().After(expiresAt) {
+			return "", 0, err
+		}
+	}
+	return "",
+		0,
+		errors.New("rpk bug, please describe how you encountered this at https://github.com/redpanda-data/redpanda/issues/new?assignees=&labels=kind%2Fbug&template=01_bug_report.md")
+}
+
+// deviceAuthorizationResponse is a response for an OAuth 2 device
+// authorization request. The struct follows the RFC8628 definition, for
+// documentation on fields, refer to the following link:
+//
+//	https://datatracker.ietf.org/doc/html/rfc8628#section-3.2
+type deviceAuthorizationResponse struct {
+	DeviceCode              string `json:"device_code"`
+	UserCode                string `json:"user_code"`
+	VerificationURI         string `json:"verification_uri"`
+	VerificationURIComplete string `json:"verification_uri_complete"`
+	ExpiresIn               int    `json:"expires_in"`
+	Interval                int    `json:"interval"`
+}
+
+func (cl *Client) getDeviceCode(ctx context.Context) (*deviceAuthorizationResponse, error) {
+	path := httpapi.Pathfmt("%s/oauth/device/code", cl.endpoint.URL)
+	form := httpapi.Values(
+		"client_id", cl.endpoint.ClientID,
+		"audience", cl.endpoint.Audience,
+	)
+
+	var response *deviceAuthorizationResponse
+	return response, cl.httpCl.PostForm(ctx, path, nil, form, response)
+}
+
+// tokenResponse is a response for an OAuth 2 access token request. The struct
+// follows the RFC6749 definition, for documentation on fields, see sections
+// 4.2.2 and 4.2.2.1:
+//
+//	https://datatracker.ietf.org/doc/html/rfc6749#section-4.2.2
+//
+// For a higher-level description, see:
+//
+//	https://www.oauth.com/oauth2-servers/access-tokens/access-token-response/
+type tokenResponse struct { //nolint:errname // This is used as an error once below, but primarily it is not an error.
+	AccessToken string `json:"access_token"`
+	TokenType   string `json:"token_type"`
+	ExpiresIn   int    `json:"expires_in"`
+	Scope       string `json:"scope"`
+
+	Err              string `json:"error"`
+	ErrorDescription string `json:"error_description"`
+}
+
+func (t *tokenResponse) Error() string {
+	if t.ErrorDescription != "" {
+		return t.Err + ": " + t.ErrorDescription
+	}
+	return t.Err
+}
+
+func (cl *Client) getToken(ctx context.Context, deviceCode string) (*tokenResponse, error) {
+	path := httpapi.Pathfmt("%s/oauth/token", cl.endpoint.URL)
+	form := httpapi.Values(
+		"grant_type", "urn:ietf:params:oauth:grant-type:device_code",
+		"device_code", deviceCode,
+		"client_id", cl.endpoint.ClientID,
+	)
+
+	var response *tokenResponse
+	return response, cl.httpCl.PostForm(ctx, path, nil, form, response)
```

Comment:
In the previous version, `resp` was returned if the endpoint returned 4xx. This new version only returns an error if the POST fails.

How about we:
* split Err and ErrorDescription fields into a separate type
* use httpapi.Err4xx to return this custom type

Then I _think_ this line can stay the same

Reply:
Change: https://github.com/redpanda-data/redpanda/compare/53781d7a8fb09c84ce2333f54376bf4208a320d2..47aa3b307f992863cac2aaec024c4629ede08258